tests/test_solving.py

```diff
@@ -170,3 +173,75 @@ def test_inconsistent_entry_number():
         (array('I', [2, 1]), array('I', [6])))
     with pytest.raises(ValueError):
         greedy_solve(candidates)
+
+
+# === HYPOTHESIS TESTS ===
+
+def dict_to_candidate_pairs(candidate_dict):
+    candidates = map(tuple, map(reversed, candidate_dict.items()))
+    return sorted(candidates, key=lambda x: (-x[0],) + x[1:])
+
+
+indices_np = strategies.tuples(
+    strategies.integers(min_value=0),
+    strategies.integers(min_value=0))
+index_pair_np = strategies.tuples(
+        indices_np, indices_np
+    ).filter(
+        lambda x: x[0] != x[1]
+    ).map(lambda x: tuple(sorted(x)))
+candidate_pairs_np = strategies.dictionaries(
+        index_pair_np,
+        strategies.floats(min_value=0, max_value=1)
```

Comment:
Probably a good time to check what happens when these bounds are not adhered to.

Reply:
The similarity scores are completely ignored in the solver right now: it only cares about the order of the candidate pairs. But you raise a valid point.

Also, at some point I will write up all the assumptions I made about the candidate pairs type.

Reply:
I think I would like to put similarities beyond those bounds under _undefined behaviour_ and just merge this PR.